fontforgeexe/pythonui.c

```diff
@@ -50,41 +51,223 @@
+ *
+ * And, of course, if we *start* with "Annuler (_C)" or "キャンセル(_C)" and
+ * "C" is not available we don't want to accidentally do
+ *
+ * mmn[j].ti.text              = "Annuler (C) (4)";
+ * mmn[j].ti.text_untranslated = "Annuler (C)";
+ * mmn[j].ti.mnemonic          = '4';
+ *
+ * so we need to strip off the parenthetical when reassigning (and hope that's
+ * the convention the translator is using).
+ *
+ * ti.text is a (unichar_t *) but can be assigned a utf8 string when setting
+ * ti.text_is_1byte to true. ti.text_untranslated is a (char *) that we just
+ * have to hope can handle non-ASCII UTF-8 content, because that may be all we
+ * have.
+ */
+
+static const char mn_order[] = "}{:[]&=~%^;?+/,#$-@*!.7869054321QXZVJWFYGKUBHPCDMLTNSIROEA";
+
+struct py_menu_item {
     PyObject *func;
-    PyObject *check_enabled;		/* May be None (which I change to NULL) */
+    PyObject *check;		/* May be None (which I change to NULL) */
     PyObject *data;			/* May be None (left as None) */
-} *cvpy_menu_data = NULL, *fvpy_menu_data = NULL;
-static int cvpy_menu_cnt=0, cvpy_menu_max = 0;
-static int fvpy_menu_cnt=0, fvpy_menu_max = 0;
+};
+
+typedef void (*ff_menu_callback)(GWindow gw, struct gmenuitem *mi, GEvent *e);
+
+enum py_menu_type { pmt_font=0, pmt_char=1, pmt_size=2 };
+enum py_menu_flag { pmf_font=1, pmf_char=2 };
+
+static struct py_menu_data {
+    struct py_menu_item *items;
+    int cnt, max;
+    GMenuItem2 *menu;
+    ff_menu_callback moveto, invoke;
+    void (*setmenu)(GMenuItem2 *menu);
+    uint8 mn_offset;
+    uint8 mn_avail[128];
+    char *hotkey_prefix;
+} *py_menus;
 
-GMenuItem2 *cvpy_menu, *fvpy_menu;
+static unichar_t AllocateNextMnemonic(struct py_menu_data *pmd) {
+    while ( *(mn_order+pmd->mn_offset)!=0 && !pmd->mn_avail[*(mn_order + pmd->mn_offset)] )
+	++pmd->mn_offset;
 
-static void py_tllistcheck(struct gmenuitem *mi,PyObject *owner,
-	struct python_menu_info *menu_data, int menu_cnt) {
+    char r = *(mn_order+pmd->mn_offset);
+    if ( r!=0 )
+	pmd->mn_avail[r] = false;
+
+    return (unichar_t) r;
+}
+
+static unichar_t FindMnemonic(const char *menu_string, unichar_t alt, struct py_menu_data *pmd) {
+    uint8 *cp;
+
+    if ( pmd->mn_avail[alt] ) {
+	pmd->mn_avail[alt] = false;
+	return alt;
+    }
+    for ( cp = (uint8 *) menu_string; *cp!=0; ++cp ) {
+	if ( *cp > 127 )
+	    continue;
+	alt = (unichar_t) toupper(*cp);
+	if ( pmd->mn_avail[alt] ) {
+	    pmd->mn_avail[alt] = false;
+	    return alt;
+	}
+    }
+    return 0;
+}
+
+static char *mncopy(const char *str, unichar_t *mn, int extra) {
+    unichar_t t=0;
+    const char *src = str;
+    char *dst, *ret;
+    size_t l=0;
+
+    if ( str==NULL ) {
+	if ( mn!=NULL )
+	    *mn = 0;
+	ret = malloc(1+extra);
+	ret[0] = 0;
+	return ret;
+    }
+
+    while ( *src!=0 ) {
+	if ( *src!='_' )
+	    ++l;
+	++src;
+    }
+
+    src = str;
+    ret = dst = malloc(l+1+extra);
+    for( ; *src; src++ ) {
+        if( *src == '_' ) {
+	    t = *(src+1);
```

Comment:
This is going to break if the following character is not ascii. Also noting the lack of the use of `utf82u_mncopy`